### PR TITLE
`ruff server` default to current working directory in environments without any root directories or workspaces

### LINKED ISSUE #10324: Server fails to start if there's not root directory provided
(Posting from Discord for tracking purposes.)

The server fails to start if there's no root directory provided in the initialization parameters:

```
ruff failed
  Cause: No workspace or root URI was given in the LSP initialization parameters. The server cannot start.
```

We default to the current working directory if that's not provided in `ruff-lsp`: https://github.com/astral-sh/ruff-lsp/blob/c410d6b44c1b40a41d254d13e33535bd198935e0/ruff_lsp/server.py#L1656-L1664

/cc @snowsignal

### PR DESCRIPTION
**Pull request by snowsignal**
## Summary

Fixes #10324

This removes an overeager failure case where we would exit early if no root directory or workspace folders were provided on server initialization. We now fall-back to the current working directory as a workspace for that file.

## Test Plan
N/A

### CODE REVIEW

**Comment by dhruvmanila** [comment]
Testing this out...

File: crates/ruff_server/src/server.rs [hunk 1]
```diff
@@ -46,9 +45,7 @@ impl Server {
             .workspace_folders
             .map(|folders| folders.into_iter().map(|folder| folder.uri).collect())
             .or_else(|| init_params.root_uri.map(|u| vec![u]))
-            .ok_or_else(|| {
-                anyhow!("No workspace or root URI was given in the LSP initialization parameters. The server cannot start.")
-            })?;
+            .unwrap_or_default();
```
Comment (dhruvmanila) [inline crates/ruff_server/src/server.rs:48]: If I understand this correctly, then the workspaces will be empty, right? Should we instead default to using the current working directory? In `ruff-lsp`, we default to using the current working directory as the workspace with the default settings. You might be more aware of how that will affect other parts of the code.

**Comment by dhruvmanila** [comment]
It's not giving me any diagnostics. The LSP logs have this error:

```
[ERROR][2024-03-14 09:56:07] .../vim/lsp/rpc.lua:789	"rpc"	"/Users/dhruv/work/astral/ruff-test/target/release/ruff"	"stderr"	"┐ruff_server::server::api::notifications::did_open::run{file=file:///Users/dhruv/playground/python/animation.py}\n┘\n"
```

I suspect this is because there's no workspace associated with that document, so the `did_open` notification is failing?

File: crates/ruff_server/src/server.rs (hunk 1 above)
Comment (dhruvmanila) [reply crates/ruff_server/src/server.rs:48]: I think we should use the current working directory as the default workspace if not provided.

Comment (snowsignal) [reply crates/ruff_server/src/server.rs]: You're right, we should try to use the current working directory. Let me know if that works!

**Comment by snowsignal** [comment]
@dhruvmanila Right, it's failing to open because the file isn't associated with any workspace. The assumption that the server has made up until now is that any file being opened by the editor would have an associated workspace folder the client would have previously told us about, but this is clearly not the case.

**Comment by snowsignal** [comment]
Actually, maybe what we need to do is just always create a new workspace in the session as a fallback mechanism if an file being opened isn't in a workspace path already.

**Review by dhruvmanila — APPROVED** [review]
Thank you! I tested out using diagnostics, code actions, formatting and range formatting. It works well. LGTM!

**Pull request merged**